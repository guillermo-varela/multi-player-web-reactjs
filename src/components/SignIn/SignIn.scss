@media all and (min-width: 480px) {
  .loginForm {
    margin: 0 auto;
    max-width: 350px;
  }
}
