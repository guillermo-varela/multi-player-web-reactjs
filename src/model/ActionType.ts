enum ActionType {
  SIGN_IN_PENDING,
  SIGN_IN_SUCCESS,
  SIGN_IN_ERROR,
  SIGN_OUT_PENDING,
  SIGN_OUT_SUCCESS,
  SIGN_OUT_ERROR,
  CHECK_AUTHENTICATION_PENDING,
  CHECK_AUTHENTICATION_PENDING_TOO_LONG,
  CHANGE_LOCALE
}

export default ActionType;
