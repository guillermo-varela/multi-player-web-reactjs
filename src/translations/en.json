{
  "component.about.code": "The source code for this project can be found at:",
  "component.about.codeSplit": "Route based code splitting",
  "component.about.csrf": "Use of custom HTTP headers in stateless/sessionless environments to prevent CSRF attacks",
  "component.about.forms.formik": "Forms in React using Formik",
  "component.about.hello": "Hi again :)",
  "component.about.internationalization": "Support for multiple languages (internationalization) using react-intl",
  "component.about.introduction": "This web application was developed to learn about:",
  "component.about.jwt": "Stateless/sessionless authentication on a server using \"Javascript Object Signing and Encryption\" (JOSE) and \"JSON Web Token\" (JWT)\"",
  "component.about.openAPI": "OpenAPI 3",
  "component.about.pixabay": "The multimedia content was fetched using the {link}.",
  "component.about.pixabay.name": "Pixabay API",
  "component.about.redux.typescript": "Redux using TypeScript",
  "component.about.rest.webflux": "REST services with reactive architecture (non-blocking) with Spring WebFlux and Spring Security",
  "component.error.message": "Ops! An error occurred...",
  "component.signin.button.sending": "Sending...",
  "component.signin.button.submit": "Sign In",
  "component.signin.credentials.hint": "Use user@test.com as email and password to test",
  "component.signin.empty.email": "Enter an email",
  "component.signin.incorrect.credentials": "Incorrect credentials",
  "component.signin.invalid.email": "Enter a valid email",
  "component.signin.invalid.password": "Enter a password",
  "component.signin.label.email": "Email",
  "component.signin.label.password": "Password",
  "menu.about": "About",
  "menu.home": "Home",
  "menu.images": "Images",
  "menu.more": "View More",
  "menu.sections": "Sections",
  "menu.signIn": "Sign In",
  "menu.signOut": "Sign Out",
  "menu.signOut.error.message": "Sign out failed",
  "menu.signOut.error.title": "Ops!",
  "menu.videos": "Videos",
  "multimedia.source": "Content from ",
  "multimedia.type.image": "Image",
  "multimedia.type.video": "Video",
  "nature.hello": "Hello!",
  "nature.introduction": "On this site you can find multimedia content about nature.",
  "nature.multimedia.endOfContent": "That's all for now...",
  "nature.title": "Nature",
  "root.loading": "Loading...",
  "root.locale.label": "Language",
  "root.nature.description": "Images and videos of flora and fauna.",
  "root.space.description": "Images and videos of the universe.",
  "space.hello": "Hello!",
  "space.images.introduction": "Multiple images related to outer space.",
  "space.introduction": "On this site you will find multimedia content about outer space.",
  "space.menu.title": "Space",
  "space.title": "Outer Space",
  "space.videos.introduction": "Mind-blowing videos from outer space."
}
