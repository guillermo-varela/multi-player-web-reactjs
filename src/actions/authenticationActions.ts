import ThunkResult from '../model/ThunkResult';
import Action from '../model/Action';
import ActionType from '../model/ActionType';
import { signIn, signOut, checkAuthentication } from '../services/authenticationService';

export const signInAction = (formData: FormData): ThunkResult => (dispatch) => {
  dispatch({ type: ActionType.SIGN_IN_PENDING });

  return signIn(formData)
    .then(user => dispatch({ type: ActionType.SIGN_IN_SUCCESS, payload: user }))
    .catch(error => dispatch({ type: ActionType.SIGN_IN_ERROR, payload: error }));
}

export const checkAuthenticationAction = (): ThunkResult => (dispatch) => {
  dispatch({ type: ActionType.CHECK_AUTHENTICATION_PENDING });

  return checkAuthentication()
    .then(user => dispatch({ type: ActionType.SIGN_IN_SUCCESS, payload: user }))
    .catch(() => dispatch({ type: ActionType.SIGN_OUT_SUCCESS }));
}

export const signOutAction = (): ThunkResult => (dispatch) => {
  dispatch({ type: ActionType.SIGN_OUT_PENDING });

  return signOut()
    .then(() => dispatch({ type: ActionType.SIGN_OUT_SUCCESS }))
    .catch(() => dispatch({ type: ActionType.SIGN_OUT_ERROR }));
}

export const simpleSignOutAction = (): Action => {
  return { type: ActionType.SIGN_OUT_SUCCESS };
}
