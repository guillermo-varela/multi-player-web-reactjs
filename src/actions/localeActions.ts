import Action from '../model/Action';
import ActionType from '../model/ActionType';
import { saveLocale } from '../services/localeService';

export const changeLocaleAction = (locale: string): Action => {
  return { type: ActionType.CHANGE_LOCALE, payload: saveLocale(locale) };
}
