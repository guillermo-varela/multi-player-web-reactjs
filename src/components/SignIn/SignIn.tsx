import React from 'react';
import { useLocation, Redirect } from 'react-router-dom';
import Form from 'react-bootstrap/Form';
import Button, { ButtonProps } from 'react-bootstrap/Button';
import Spinner from 'react-bootstrap/Spinner';
import { Formik } from 'formik';
import * as Yup from 'yup';
import { useDispatch } from 'react-redux';
import { FormattedMessage, useIntl } from 'react-intl';
import ErrorAlert from '../../components/ErrorAlert/ErrorAlert';
import { signInAction } from '../../actions/authenticationActions';
import useTypedSelector from '../../hooks/useTypedSelector';

import './SignIn.scss';

type Variant = {
  variant: ButtonProps['variant']
}

const SignIn: React.FC<Variant> = ({ variant }) => {
  const authenticationState = useTypedSelector(state => state.authentication);
  const intl = useIntl();
  const dispatch = useDispatch();
  const location = useLocation();
  const { referrer } = (location.state as any) || {
    referrer: {
      pathname: location.pathname.substring(0, location.pathname.lastIndexOf('/'))
    }
  };

  if (authenticationState.user) {
    return <Redirect to={referrer} />
  }

  return (
    <Formik
      initialValues={{
        username: 'user@test.com',
        password: 'user@test.com'
      }}
      validationSchema={Yup.object({
        username: Yup.string()
          .email(intl.formatMessage({ id: 'component.signin.invalid.email' }))
          .required(intl.formatMessage({ id: 'component.signin.empty.email' })),
        password: Yup.string()
          .required(intl.formatMessage({ id: 'component.signin.invalid.password' }))
      })}
      onSubmit={async (values) => {
        const formData = new FormData();
        Object.entries(values).forEach(([key, value]) => formData.append(key, value));
        return dispatch(signInAction(formData));
      }}
    >
      {formik => (
        <Form noValidate onSubmit={formik.handleSubmit} className="loginForm">
          <Form.Group controlId="formUsername">
            <Form.Label>
              <FormattedMessage id='component.signin.label.email' />
            </Form.Label>
            <Form.Control type="email" name="username" autoFocus
                          isValid={formik.touched.username && !formik.errors.username}
                          isInvalid={formik.touched.username && !!formik.errors.username}
                          {...formik.getFieldProps('username')} />
            <Form.Control.Feedback type="invalid">
              {formik.errors.username}
            </Form.Control.Feedback>
            <Form.Text className="text-muted">
              <FormattedMessage id='component.signin.credentials.hint' />
            </Form.Text>
          </Form.Group>

          <Form.Group controlId="formPassword">
            <Form.Label>
              <FormattedMessage id='component.signin.label.password' />
            </Form.Label>
            <Form.Control type="password" name="password"
                          isValid={formik.touched.password && !formik.errors.password}
                          isInvalid={formik.touched.password && !!formik.errors.password}
                          {...formik.getFieldProps('password')} />
            <Form.Control.Feedback type="invalid">
              {formik.errors.password}
            </Form.Control.Feedback>
          </Form.Group>

          <Button variant={variant} type="submit" size="lg"
                  disabled={formik.isSubmitting || authenticationState.isPendingSignIn} block>
                  {formik.isSubmitting || authenticationState.isPendingSignIn ?
              <div>
                <Spinner as="span" animation="border" size="sm" role="status" aria-hidden="true" />
                &nbsp;<FormattedMessage id='component.signin.button.sending' />
              </div>
              : <FormattedMessage id='component.signin.button.submit' />
            }
          </Button>

          {authenticationState.isErrorSignIn &&
            <div>
              <br />
              <ErrorAlert message={authenticationState.isBadCredentials ?
                intl.formatMessage({ id: 'component.signin.incorrect.credentials' })
                : ''} />
            </div>
          }
        </Form>
      )}
    </Formik>
  );
}

export default SignIn;
