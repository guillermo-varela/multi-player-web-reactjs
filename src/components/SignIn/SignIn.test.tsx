import React from 'react';
import { act } from 'react-dom/test-utils';
import { MemoryRouter, Redirect } from 'react-router-dom';
import { Provider } from 'react-redux';
import thunk from 'redux-thunk';
import configureMockStore, { MockStoreEnhanced } from 'redux-mock-store';
import { mount, ReactWrapper } from 'enzyme';
import renderer from 'react-test-renderer';
import { IntlProvider } from 'react-intl';
import en from '../../translations/en.json';
import { signInAction } from '../../actions/authenticationActions';
import ActionType from '../../model/ActionType';
import SignIn from './SignIn';

jest.mock('react-router-dom', () => ({
  ...jest.requireActual('react-router-dom'),
  useLocation: jest.fn()
}));

jest.mock('../../actions/authenticationActions');
const signInActionMock = signInAction as jest.Mock;

import { useLocation } from 'react-router-dom';
const useLocationMock = useLocation as jest.Mock;

describe('<SignIn />', () => {
  const mockStore = configureMockStore([thunk]);
  let store: MockStoreEnhanced<unknown, {}>;
  const defaultUsername = 'user@test.com';

  beforeEach(() => {
    signInActionMock.mockReset();
    signInActionMock.mockReturnValue({ type: ActionType.SIGN_IN_PENDING });
  });

  it('renders <Redirect /> to parent path when authenticated', () => {
    const wrapper = mount(buildComponent(true));
    const redirect = wrapper.find(Redirect);
    expect(redirect.length).toBe(1);
    expect(redirect.prop('to')).toEqual({ pathname: '/test' });
  });

  it('renders <Redirect /> to a given referrer path after successful sign in', () => {
    const wrapper = mount(buildComponent(true, false, false, false, {
      state: {
        referrer: {
          pathname: '/test/sign-in-required'
        }
      }
    }));
    const redirect = wrapper.find(Redirect);
    expect(redirect.length).toBe(1);
    expect(redirect.prop('to')).toEqual({ pathname: '/test/sign-in-required' });
  });

  it('renders a form with initial values when not authenticated', () => {
    const wrapper = mount(buildComponent());
    const form = wrapper.find('form');
    expect(form.length).toBe(1);

    const usernameInput = form.find('input[name="username"]');
    expect(usernameInput.prop('value')).toBe(defaultUsername);

    const passwordInput = form.find('input[name="password"]');
    expect(passwordInput.prop('value')).toBe(defaultUsername);

    const submitButton = form.find('Button');
    expect(submitButton.prop('variant')).toBe('warning');
    expect(submitButton.prop('disabled')).toBe(false);
    expect(submitButton.find('Spinner').length).toBe(0);
  });

  it('submits a sign in action with the form data', async () => {
    const wrapper = mount(buildComponent());
    await act(async () => {
      wrapper.find('form').simulate('submit');
    });
    const formData = new FormData();
    formData.append('username', defaultUsername);
    formData.append('password', defaultUsername);
    expect(signInActionMock).toHaveBeenCalledWith(formData);
  });

  it('disables submit button while signing in', async () => {
    const wrapper = mount(buildComponent(false, true, false, false));
    const submitButton = wrapper.find('button[type="submit"]');
    expect(submitButton.prop('disabled')).toBe(true);
    expect(submitButton.find('Spinner').length).toBe(1);
  });

  it('renders <ErrorAlert /> with default message when sign in fails', async () => {
    const wrapper = mount(buildComponent(false, false, true, false));
    const errorAlert = wrapper.find('ErrorAlert');
    expect(errorAlert.length).toBe(1);
    expect(errorAlert.prop('message')).toBeFalsy();
  });

  it('renders <ErrorAlert /> with custom message for bad credentials', async () => {
    const wrapper = mount(buildComponent(false, false, true, true));
    const errorAlert = wrapper.find('ErrorAlert');
    expect(errorAlert.length).toBe(1);
    expect(errorAlert.prop('message')).toBeTruthy();
  });

  it('disables form submission when invalid data is entered', async () => {
    const wrapper = mount(buildComponent());
    const form = wrapper.find('form');
    await updateField(form, 'username', '');
    await submitForm(form);

    const usernameFormGroup = form.find('FormGroup[controlId="formUsername"]');
    expect(usernameFormGroup.find('div.invalid-feedback').text()).toBeTruthy()
    expect(signInActionMock).toHaveBeenCalledTimes(0);

    await updateField(form, 'username', defaultUsername);
    await updateField(form, 'password', '');
    await submitForm(form);

    const passwordFormGroup = form.find('FormGroup[controlId="formPassword"]');
    expect(usernameFormGroup.find('div.invalid-feedback').text()).toBeFalsy();
    expect(passwordFormGroup.find('div.invalid-feedback').text()).toBeTruthy();
    expect(signInActionMock).toHaveBeenCalledTimes(0);
  });

  it('matches snapshot', () => {
    const tree = renderer.create(buildComponent()).toJSON();
    expect(tree).toMatchSnapshot();
  });

  const buildComponent = (
    isAuthenticated: boolean = false,
    isPendingSignIn: boolean = false,
    isErrorSignIn: boolean = false,
    isBadCredentials: boolean = false,
    location: any = { pathname: '/test/signIn' }): JSX.Element => {
    useLocationMock.mockReturnValueOnce(location);
    const authentication: any = isAuthenticated ? { user: {} } : {};
    authentication.isPendingSignIn = isPendingSignIn;
    authentication.isErrorSignIn = isErrorSignIn;
    authentication.isBadCredentials = isBadCredentials;
    store = mockStore({ authentication: authentication });
    return (
      <Provider store={store}>
        <MemoryRouter>
          <IntlProvider locale='en' messages={en}>
            <SignIn variant="warning" />
          </IntlProvider>
        </MemoryRouter>
      </Provider>
    );
  }

  const updateField = async (formWrapper: ReactWrapper, fieldName: string, value: string) => {
    await act(async () => {
      formWrapper.find(`input[name="${fieldName}"]`).simulate('change', {
        target: {
          name: fieldName,
          value: value
        }
      });
    });
  }

  const submitForm = async (formWrapper: ReactWrapper) => {
    await act(async () => {
      formWrapper.find('button[type="submit"]').simulate(
        'submit',
        { preventDefault: () => { } }
      );
    });
  }
});
