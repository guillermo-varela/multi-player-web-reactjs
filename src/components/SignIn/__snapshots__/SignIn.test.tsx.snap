// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SignIn /> matches snapshot 1`] = `
<form
  className="loginForm"
  noValidate={true}
  onSubmit={[Function]}
>
  <div
    className="form-group"
  >
    <label
      className="form-label"
      htmlFor="formUsername"
    >
      Email
    </label>
    <input
      autoFocus={true}
      className="form-control"
      id="formUsername"
      name="username"
      onBlur={[Function]}
      onChange={[Function]}
      type="email"
      value="user@test.com"
    />
    <div
      className="invalid-feedback"
    />
    <small
      className="text-muted form-text"
    >
      Use user@test.com as email and password to test
    </small>
  </div>
  <div
    className="form-group"
  >
    <label
      className="form-label"
      htmlFor="formPassword"
    >
      Password
    </label>
    <input
      className="form-control"
      id="formPassword"
      name="password"
      onBlur={[Function]}
      onChange={[Function]}
      type="password"
      value="user@test.com"
    />
    <div
      className="invalid-feedback"
    />
  </div>
  <button
    className="btn btn-warning btn-block btn-lg"
    disabled={false}
    type="submit"
  >
    Sign In
  </button>
</form>
`;
